src/Compilers/CSharp/Portable/Symbols/Metadata/PE/TupleTypeDecoder.cs

```diff
@@ -0,0 +1,313 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using Roslyn.Utilities;
+using System;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using System.Reflection.Metadata;
+
+namespace Microsoft.CodeAnalysis.CSharp.Symbols.Metadata.PE
+{
+    /// <summary>
+    /// In C#, tuples can be represented using tuple syntax and be given
+    /// names. However, the underlying representation for tuples unifies
+    /// to a single underlying tuple type, System.ValueTuple. Since the
+    /// names aren't part of the underlying tuple type they have to be
+    /// recorded somewhere else.
+    /// 
+    /// Roslyn records tuple names in an attribute: the
+    /// TupleElementNamesAttribute. The attribute contains a single string
+    /// array which records the names of the tuple elements in a breadth-first
+    /// traversal. If the type contains nested parameters, they are also recorded
+    /// breadth-first. <see cref="DecodeTupleTypesIfApplicable(TypeSymbol, EntityHandle, PEModuleSymbol)"/>
+    /// can be used to extract tuple names and types from metadata and create
+    /// a <see cref="TupleTypeSymbol"/> with attached names.
+    /// 
+    /// <example>
+    /// For instance, a method returning a tuple
+    /// 
+    /// <code>
+    ///     (int x, int y) M() { ... }
+    /// </code>
+    ///
+    /// will be encoded using an attribute on the return type as follows
+    /// 
+    /// <code>
+    ///     [return: TupleElementNamesAttribute(new[] { "x", "y" })]
+    ///     System.ValueTuple&lt;int, int&gt; M() { ... }
+    /// </code>
+    /// </example>
+    /// 
+    /// <example>
+    /// For nested type parameters, we expand the tuple names in breadth-first
+    /// order:
+    /// 
+    /// <code>
+    ///     class C : BaseType&lt;((int e1, int e2) e3, int e4)&lt; { ... }
+    /// </code>
+    ///
+    /// becomes
+    /// 
+    /// <code>
+    ///     [TupleElementNamesAttribute(new[] { "e3", "e4", "e1", "e2" });
+    ///     class C : BaseType&lt;System.ValueTuple&lt;
+    ///         System.ValueTuple&lt;int,int&gt;, int&gt;
+    ///     { ... }
+    /// </code>
+    /// </example>
+    /// </summary>
+    internal struct TupleTypeDecoder
```

Comment:
Rather than a common base class, perhaps the traversal could be a simple `TransformType` extension method similar to `TypeSymbolExtensions.VisitType`.